Edit comprehend entities and keep analysis pane open during and after transcription

ShivaniMehendarge opened the pull request:
*Issue #, if available:*

*Description of changes:*
1. Allow adding and deleting of comprehend entities
2. Keep analysis pane open during and after transcription
3. Allow editing transcription after the audio/recording is complete

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

ShivaniMehendarge committed:
remove heightslider and introduce collapse components for icd10cm and rx
norm concepts

Commit:
progress

Commit:
change comprehend result mechanism

Commit:
entity deletion

Commit:
sort descending

ShivaniMehendarge committed:
display confidence scores and types for entities and make transcriptpane editable

Commit:
fix transcript editing

ShivaniMehendarge committed:
Add functionality to add comprehend entities on analysis pane

ShivaniMehendarge committed:
This commit includes following changes:
1. Allow editing transcription only after the recording is complete.
2. Make analysis pane visible during and after transcription
3. Change the delete icon used for deleting entities
4. Allow adding only non empty strings as entities on the analysis pane.

ShivaniMehendarge committed:
add deleteicon

alexchirayath commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
-  }, [ result ]);
+  }, [result]);
+
+  if (!result.ICD10CMConcepts && !result.RxNormConcepts && !result.Attributes) {
```
Can both these if conditions be merged together or refactored to avoid duplication?
```
If (!ICD && !RXnorm){
if (result.Attr)
else {
}

}
```

kazbaig replied:
To add to this, moving the expression in the `if` statements to variables would help understand why the condition results in the returned value

kazbaig commented on app/package.json:
```diff
@@ -107,6 +112,9 @@
     "websocket-extensions": "^0.1.4",
     "yargs-parser": "^18.1.3"
   },
+  "resolutions": {
```
Why was this added?

ShivaniMehendarge replied:
It was not supposed to be added. Will remove.

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+      <Select
+        mr={2}
+        border='1px solid'
+        borderColor={concepts[0].Score < 0.5 ? 'red' : 'grey'}
```
Can extract `concepts[0].Score < 0.5 ? 'red' : 'grey'` to a variable for reuse below

alexchirayath replied:
To add on to that,  you could think of moving the confidence score threshold to a constant value. We may want to enable the user in the future to enter their custom defined threshold

kazbaig commented on app/src/useComprehension.js:
```diff
@@ -19,42 +19,42 @@ const resultMap = new WeakMap();
 // React hook to take an array of transcript chunks and return a corresponding array of Comprehend results, one for each
 export default function useComprehension(transcriptChunks, clientParams) {
   // Dummy state variable so we can force a refresh when we receive an update
```
This comment should be removed as it does not apply to the new state

ShivaniMehendarge replied:
Will do.

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+          flex='1'
+          mr={2}
+          height='2.5rem'
+          border={result.Score && result.Score < 0.5 ? '1px solid red' : '1px solid grey'}
```
Edge case - it will be grey instead of red when `result.Score` is `0` because `0` is falsy. Is that okay with us? Will we ever receive a score of `0`?

ShivaniMehendarge replied:
I haven't encountered any instance where the score is 0. the model should not detect the entity if it has 0 confidence in it ideally right?

kazbaig replied:
Probably. Just want to make sure we've considered it first.

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+            placeholder={'Add ' + displayNames[category]}
+            value={inputValue}
+            onChange={handleInputChange}
+          ></Input>
```
Components/elements not receiving a `children` prop should be self-closing in JSX.

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+            borderColor='grey'
+            borderRadius='0'
+            bg='white'
+            placeholder={'Add ' + displayNames[category]}
```
Template syntax is idiomatically preferred over string concatenation

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+        <Flex width='100%' alignItems='center'>
+          <Input
+            mr={2}
+            mb={4}
```
This margin is causing misalignment with the button. It should go on their parent.

![Screen Shot 2021-03-08 at 2 21 40 PM](https://user-images.githubusercontent.com/72631444/110377826-7972be00-801a-11eb-891e-4209bf8e1724.png)

ShivaniMehendarge replied:
Got it. Thanks!

kazbaig commented on app/src/components/AnalysisPane.module.css:
```diff
@@ -54,7 +50,8 @@
 }
 
 .resultTable {
```
Since the user of this class is now a chakra component, let's use style props to remove this css declaration rather than mixing style props and css.

ShivaniMehendarge replied:
Will do.

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+  const [inputValue, setInputValue] = React.useState('');
+  const handleInputChange = (event) => setInputValue(event.target.value);
+
+  const addIcon = (
```
The hover state is really difficult to see against the background (bad for accessibility). Let's work with design to improve it.

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+        >
+          {attrs.map(([key, value]) => (
+            <React.Fragment key={key}>
+              {result.Text} {value ? '|' : null} {value}
```
`{ booleanyValue && <Component /> }` is preferred for conditional rendering

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+    );
+  }
 
+  const concepts = [...(result.ICD10CMConcepts ? result.ICD10CMConcepts : result.RxNormConcepts)];
```
IMO this is more concise
```js
const concepts = [...(result.ICD10CMConcepts ? result.ICD10CMConcepts : result.RxNormConcepts)].sort(
    (concept1, concept2) => concept2.Score - concept1.Score,
  );
```

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+      >
+        {concepts.map((concept) => (
+          <option key={concept.Code} value={concept.Code}>
+            {result.Text} {concept.Code ? ' | ' : null} {concept.Code} {concept.Description ? ' | ' : null}{' '}
```
Same comment about conditional rendering

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+        bg='white'
+        _hover={{ borderColor: concepts[0].Score < 0.5 ? 'red' : 'grey', boxShadow: 'none' }}
+      >
+        {concepts.map((concept) => (
```
Destructuring the `(concept) =>` to `({ Code, Text, Description }) => ` should make this a bit cleaner

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+        {concepts.map((concept) => (
+          <option key={concept.Code} value={concept.Code}>
+            {result.Text} {concept.Code ? ' | ' : null} {concept.Code} {concept.Description ? ' | ' : null}{' '}
+            {concept.Description} &nbsp;&nbsp; {(concept.Score * 100).toPrecision(4)}%
```
Is `concept.Score` potentially undefined? `undefined * 100` will be `NaN`.

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+    <IconButton
+      aria-label='Add'
+      icon={<AddIcon />}
+      onClick={() => [inputValue.trim() !== '' ? onResultAdd(inputValue, category) : null, setInputValue('')]}
```
Why is this an array? Looks like it should be a function that calls 2 functions.

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+    <IconButton
+      aria-label='Add'
+      icon={<AddIcon />}
+      onClick={() => [inputValue.trim() !== '' ? onResultAdd(inputValue, category) : null, setInputValue('')]}
```
We should call `onResultAdd` with the trimmed value as a better UX. Also, the input should only be cleared when the result was added. Here, I could add a bunch of whitespace and it would cleared even if we decided not to add.

ShivaniMehendarge replied:
Good point.

kazbaig commented on app/src/components/ExportPane.js:
```diff
@@ -3,73 +3,56 @@ import React, { useMemo } from 'react';
-  result
-}) {
-
+function ResultRow({ result }) {
```
Are there any changes in this file besides the addition of the key or is it just all formatting?

ShivaniMehendarge replied:
Nope. Nothing besides those changes

kazbaig commented on app/src/home.js:
```diff
@@ -567,22 +577,43 @@ export default function Home() {
+    );
+  };
+
+  const onComprehendResultAddition = (val, category) => {
```
I would expect the addition to show as the first item in the list, right below the input element. Let's see how others feel.

ShivaniMehendarge replied:
Updated

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
+
+      <VStack spacing={2}>
+        <Flex width='100%' alignItems='center'>
+          <Input
```
[Input elements should always have labels, even if they're hidden visually.](https://www.w3.org/WAI/tutorials/forms/labels/). You can use the [`VisuallyHidden`](https://chakra-ui.com/docs/disclosure/visually-hidden) component to achieve this.

ShivaniMehendarge replied:
Added.

alexchirayath commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,194 @@
-            <dt>{key}</dt>
-            <dd>{value}</dd>
-          </>
+    <Flex width='100%' alignItems='center'>
```
If we are allowing them to choose the dropdown to select another code/item for that specific entity, is there a way to record that selection? That selection can be then used to generate the summary page. For example, can we move that selection to the top of the list for ICD10/RX norm?

kazbaig replied:
We said that we would remember the chosen concept once we decide that we're going to print it in the summary (we currently don't). Now that we have a PR open that includes that change, let's add the state piece here and use it in the summary gen PR.

ShivaniMehendarge committed:
Made the following changes:
1. Addressed most of the comments on the PR
2. Positioned the microphone icon to top left during transcription phase
3. modified the add icon for adding entities
4. added the functionality to add entities on Enter
5. Disable the add and delete entity functionality during transcription phase
6. maintain the custom entities added even after transcription is edited and the newly added entities appear at top
7. Color change to highlight entities with low confidence score and add background color to entities on Analysis Pane
8. Remove the legend that displays highlights on transcript

ShivaniMehendarge committed:
remove the resolutions from package.json

ShivaniMehendarge committed:
Remove unnecessary comment

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,233 @@
+      onClick={onDeleteClick}
+      size='xs'
+      isRound
+      isDisabled={stage === STAGE_TRANSCRIBING && true}
```
`&& true` is extra here

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,233 @@
-  }, [ onToggleItem, result.id ]);
-
-  const isExcluded = useMemo(() => excludedItems.includes(result.id), [ excludedItems, result.id ]);
+const confidenceThreshold = 0.5;
```
Let's move this to the `consts` file or a new similarly named file and use CONSTANT_CASE since this is technically a global constant

ShivaniMehendarge replied:
Good point. Made the changes, Will push in a new commit

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,233 @@
+      onClick={handleSubmit}
+      size='xs'
+      isRound
+      isDisabled={stage === STAGE_TRANSCRIBING && true}
```
`&& true` is extra here

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,147 +1,233 @@
+      <VStack spacing={2}>
+        <FormControl>
+          <Flex width='100%' mb={4} alignItems='center'>
+            <VisuallyHidden>`Add ${displayNames[category]}`</VisuallyHidden>
```
This should be `Add {displayNames[category]}`

kazbaig replied:
The rendered tag should be a `label` here, not the default `span` and should associate with the input via `htmlFor`

kazbaig commented on app/src/home.js:
```diff
@@ -565,24 +578,57 @@ export default function Home() {
 
+  const onComprehendResultDelete = (r) => {
+    setComprehendResults((prevResults) =>
+      prevResults.map((prevResult) => {
```
One problem with always returning the result of `Array.map` in a setState function is that it will always trigger a rerender, even if we didn't actually change any inner values. Let's see if there's something we actually need to change before mapping and return the previous value if not for these setState calls here.

ShivaniMehendarge replied:
As per our conversation offline, we decided to leave this as is for now.

kazbaig commented on app/src/home.js:
```diff
@@ -565,24 +578,57 @@ export default function Home() {
+  };
+
+  const onComprehendResultAddition = (val, category) => {
+    setComprehendCustomEntities((prevEntities) => {
```
IMO we should have a field to identify custom entities as such as well - something like `isCustomEntity: true` - when initializing a custom entity so that we can avoid iterating through the comp med entities when we really want to just touch the custom entities and vice versa. This will also ensure that we don't accidentally delete the wrong entity based on id collision.

ShivaniMehendarge replied:
Good point. Added that in.

kazbaig reviewed (changes requested):
Any thoughts on whether we should make the spacing even between the `input` and the entities?

![Screen Shot 2021-03-09 at 1 15 05 PM](https://user-images.githubusercontent.com/72631444/110526441-68d64c80-80db-11eb-9cb7-83958a001712.png)

ShivaniMehendarge commented:
> Any thoughts on whether we should make the spacing even between the `input` and the entities?
> 
> ![Screen Shot 2021-03-09 at 1 15 05 PM](https://user-images.githubusercontent.com/72631444/110526441-68d64c80-80db-11eb-9cb7-83958a001712.png)

Brenda had suggested that we have extra space between the input element and the entities. I don't mind having it either way.

kazbaig commented:
> > Any thoughts on whether we should make the spacing even between the `input` and the entities?
> > ![Screen Shot 2021-03-09 at 1 15 05 PM](https://user-images.githubusercontent.com/72631444/110526441-68d64c80-80db-11eb-9cb7-83958a001712.png)
> 
> Brenda had suggested that we have extra space between the input element and the entities. I don't mind having it either way.

If this was intentional, I'm good!

ShivaniMehendarge committed:
This commit
1. Addresses most PR comments
2. Fixes the issue of being able to use the enter key to add elements even during transcription
3. Add the selectedConceptCode to comprehend medical concepts in order to allow access for summary generation

alexchirayath committed:
Summary generation (#56)

* Adding speaker diarization

* Removing unnecessary console logs

* Fixing issues around adding unique key for Reactfragments

* Formatting code, removing print stmts

* Changing speaker label presentation

* Starting speaker labels from 1

* Summary gen page

* Summary generation edits

* Changes for summary + review pane

* Adding components for export and updating soap summary

* removing whistespace

* Addressing initial batch of PR comments and solving bugs introduced by rebase conflicts

* Content grammatical fixes and refactoring

* moving generate soap summary to utils

* Addressing PR comments

* Using nullish coalescing operator

* Changes for creating Header componet in Export pane

* Resolving rebase conflcit

* Remove pointer on disabled button

* conflict updates

Co-authored-by: Alex Chirayath <alexchir@3c22fb84644d.ant.amazon.com>

Commit:
Propogate custom entities and selected concepts

Commit:
Revert "Propogate custom entities and selected concepts"

This reverts commit 74c993c43f082e677b6777ac494accc7fd8d68ea.

alexchirayath reviewed (approved):
LGTM

alexchirayath committed:
Propogate custom entities (#58)

* Propogate custom entities and selected concepts

* refactoring

Co-authored-by: Alex Chirayath <alexchir@3c22fb84644d.ant.amazon.com>

kazbaig commented on app/src/components/AnalysisPane.js:
```diff
@@ -1,126 +1,247 @@
+        <FormControl as='form' onSubmit={handleSubmit}>
+          <Flex width='100%' mb={4} alignItems='center'>
+            <VisuallyHidden as='label' htmlFor={addEntityInputId}>
+              Add displayNames[category]
```
This should be an expression, not a string

ShivaniMehendarge replied:
Good catch. Fixed it.

kazbaig commented on app/src/home.js:
```diff
@@ -577,27 +601,81 @@ export default function Home() {
     stage = STAGE_EXPORT;
   }
 
+  const onComprehendResultDelete = (r) => {
```
Let's flip the boolean for the ternary and remove console logs

ShivaniMehendarge replied:
Done.

alexchirayath commented on app/src/home.js:
```diff
@@ -171,28 +173,55 @@ export default function Home() {
-      },
-      ['', true],
-    );
+    let addSpeakerLabel = true;
```
This entire codeblock needs to the one that is on master. Please revert thie addTranscriptChunk to what is on master

ShivaniMehendarge replied:
Reverted. Will push a new change

alexchirayath commented on app/src/consts.js:
```diff
@@ -4,6 +4,10 @@ export const STAGE_HOME = 0;
 export const STAGE_TRANSCRIBING = 1;
 export const STAGE_TRANSCRIBED = 2;
 export const STAGE_SUMMARIZE = 3;
```
I dont think we have a STAGE_SUMMARIZE any more.

ShivaniMehendarge replied:
Should we rename the stages below it too then?

alexchirayath reviewed (changes requested):
Comments left inline

ShivaniMehendarge committed:
Revert addTranscriptChunk changes and address comments on PR
